src/gfx/pal_packing.cpp

```diff
@@ -199,21 +200,35 @@ class AssignedProtos {
 		return colors.size() <= options.maxOpaqueColors();
 	}
 
+	// The formula we want to compute[1] is:
+	//     sum((1 / <how many of our proto-pals contain `color`>)  for `color` in `protoPal`)
+	// But, this yields a non-integer result. Using floating-point numbers would lead
+	// to imprecision, and even per-platform differences. Not good.
+	// We avoid this by multiplying the formula such that the division always produces an integer;
+	// the LCM of all values the denominator can take is the smallest suitable factor.
+	//
+	// [1]: The paper and the associated code disagree on the formula: the code adds 1 to the
+	// denominator, whereas the paper does not; its lack causes a division by 0 if the color is not
+	// found in any proto-pal, and adding it does not seem to invalidate the paper's reasoning,
+	// so it seems safe to assume the paper is wrong.
+	static constexpr uint32_t scaleFactor = 12'252'240; // LCM(1..=17)
+
 	/*
 	 * Computes the "relative size" of a proto-palette on this palette
 	 */
-	double relSizeOf(ProtoPalette const &protoPal) const {
+	uint32_t relSizeOf(ProtoPalette const &protoPal) const {
 		// NOTE: this function must not call `uniqueColors`, or one of its callers will break!
-		double relSize = 0.;
+
+		uint32_t relSize = 0;
 		for (uint16_t color : protoPal) {
 			auto n = std::count_if(RANGE(*this), [this, &color](ProtoPalAttrs const &attrs) {
 				ProtoPalette const &pal = (*_protoPals)[attrs.protoPalIndex];
 				return std::find(RANGE(pal), color) != pal.end();
 			});
-			// NOTE: The paper and the associated code disagree on this: the code has
-			// this `1 +`, whereas the paper does not; its lack causes a division by 0
-			// if the symbol is not found anywhere, so I'm assuming the paper is wrong.
-			relSize += 1. / (1 + n);
+			assume(
+			    scaleFactor % (1 + n) == 0
+			); // The scale factor should ensure integer divisions only.
```

Comment:
```suggestion
			// The scale factor should ensure integer divisions only.
			// NOTE: We add 1 to the denominator in this "multiplicity" formula, which the
			// reference code does but the paper does not. Not adding 1 causes a division by 0 if
			// the color is not found in any proto-palette, and adding it does not invalidate the
			// paper's reasoning, so it seems safe to assume the paper's formula is wrong.
			assume(scaleFactor % (1 + n) == 0);
```

Reply:
I'd slightly prefer keeping them apart, so instead of "one huge footnoted comment" it's "two comments at the two relevant sites".